### ISSUE TITLE
Refactor authentication plugin interface leveraging Authentication Working Group design

### ISSUE THREAD

**Issue by jasoncoposky** [irods/irods#5479]
A new authentication plugin interface has been created via the [AWG](https://github.com/irods-contrib/irods_working_group_authentication/tree/master/development).  This new mechanism will be ported to the main iRODS repository.

Consider backwards compatibility with older clients as well as federation.

- [x] Native
- [x] PAM
- [x] Kerberos

The following will not be carried forward into 4.3.x (see https://groups.google.com/g/iROD-Chat/c/sSlBMvI-21o):
- OS Auth
- GSI

**Comment by alanking** [comment]
Just leaving a note here for tracking progress...

- [x] Pulled all code from AWG into irods/irods
- [x] Code builds
- [x] Wired native authentication plugin into `clientLogin` so that iCommands use new authentication method

Now I'm experiencing `SYS_UNMATCHED_API_NUM` when attempting to start a catalog consumer server.

**Comment by trel** [comment]
That's... allow list?

**Comment by trel** [comment]
ah.

aka, too late.

**Comment by trel** [comment]
but also, so that it's dynamic...     hmm...

**Comment by alanking** [comment]
The issue with CSC servers has been resolved. The client API table needs to be loaded when standing up the server in order to use the authentication plugins when querying the CSP for information.

Now investigating server authentication failures across federation.

**Comment by alanking** [comment]
I broke the following unit tests as they are not loading the client API plugins:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        test list:
                [irods_client_connection]
                [irods_connection_pool]
                [irods_query_builder]
                [irods_resource_administration]
                [irods_user_administration]
                [irods_zone_report]
        failed tests:
                [irods_client_connection]
                [irods_connection_pool]
                [irods_query_builder]
                [irods_resource_administration]
                [irods_user_administration]
                [irods_zone_report]
        return code:[1]
-----
List of failed tests:
        irods_connection_pool irods_query_builder irods_resource_administration irods_user_administration irods_zone_report
Return code:[1]
==== end of test run results ====
```

**Comment by alanking** [comment]
At the most recent AWG meeting (https://github.com/irods-contrib/irods_working_group_authentication/blob/main/20220125-minutes.md), we discussed implementing the legacy PAM auth plugin using the new plugin framework. This is now done.

Presently, the PAM plugin is very closely tied to the native authentication plugin, to the point where certain database operations only used in the native authentication plugin are PAM-aware. The flow goes something like this:

1. The PAM user must run `iinit`. An initial call to `clientLogin` initiates the PAM flow wherein the PAM password is prompted and used to generate a password which is stored in the catalog. This is the only place where the PAM plugin operations are actually invoked. At this point, we are dealing only in PAM credentials and authentication with the server has not occurred. Furthermore, the `RcComm` is not marked as `loggedIn` by the plugin upon completion.
2. `iinit` continues on and authenticates with the server with a *second* `clientLogin` call using native authentication with the generated password.
3. The client is then authenticated and all other iCommands use native authentication to authenticate with the server until the password expires according to the TTL parameter, at which time the user must run `iinit` again.

There is now a fork in the road here as to how we choose to support this:

1. We keep things as they are now and the PAM plugin has to "override" the scheme provided in the environment with native.
2. The PAM authentication plugin absorbs the native authentication plugin. This can be done by simply copying and pasting the operations from the native plugin into the PAM plugin.
3. The PAM authentication plugin can attempt to initiate the native plugin flow FROM the PAM plugin (i.e. load the native plugin and call the start operation from there).

After the legacy PAM authentication plugin is completed using one of the 3 options above, we can start talking about how we would like to implement a new PAM plugin that makes more sense :)

**Comment by alanking** [comment]
From the "fork in the road" options above, I have successfully demonstrated implementations for 1 and 3. I am leaning toward option 3 at this point because I really don't like the partial authentication implementation provided by the legacy PAM plugin. Thoughts?

**Comment by trel** [comment]
3 feels... the most robust, if only because of a single source of truth for 'how to native'.

**Comment by alanking** [comment]
All of the plugins we plan to support have been implemented, at least as an initial effort. Once merged, the following things still need to be done:
 - Take a look at `iinit` and determine whether the old auth flows need to be supported
 - Consolidate common code into an authentication utilities library
   - request
   - verify
   - get?
 - Replace usage of the raw API number in `request`
 - Use const ref instead of copy of `std::vector<std::string>` in `verify`
 - Add commonly referenced strings to a header file for authentication (e.g. "user_name", "zone_name", etc.)

**Comment by alanking** [comment]
I've managed to complete the following items from the list above:
All of the plugins we plan to support have been implemented, at least as an initial effort. Once merged, the following things still 
 - Consolidate common code into an authentication utilities library
   - request
   - verify
 - Replace usage of the raw API number in `request`
 - Use const ref instead of copy of `std::vector<std::string>` in `verify`

The first bullet item about `iinit` is the most important thing to get done before closing this.

The get function is not generally useful within the plugins, so I won't be putting that in a common place. In fact, it has been replaced in all of the plugins except for the API plugin where it is defined locally. The commonly referenced strings are not very important.
